core/state/statedb.go

```diff
@@ -47,24 +47,25 @@ type revision struct {
 	journalIndex int
 }
 
-// StateDB structs within the ethereum protocol are used to store anything
-// within the merkle trie. StateDBs take care of caching and storing
-// nested states. It's the general query interface to retrieve:
-//
-// * Contracts
-// * Accounts
-//
-// Once the state is committed, tries cached in stateDB (including account
-// trie, storage tries) will no longer be functional. A new state instance
-// must be created with new root and updated database for accessing post-
-// commit states.
+// StateDB structs within the ethereum protocol are used to read, write and
+// hash ethereum states.
 type StateDB struct {
-	db         Database
+	db Database
+
+	// This reader is associated with a specific state and gets re-constructed
+	// with a new state root after a commit operation.
+	reader StateReader
```

Comment:
when the state root is reconstructed, must the root node be deserialized again?

Reply:
The associated `statedb.Trie` is reset to nil, to force a re-loading for next access.

Reply:
right, and this currently happens in any case. nvm.